### PR TITLE
Add Paraclesian imbuements

### PR DESCRIPTION
**Pull request by MorvarchPrincess**
#### Summary
Mods "Add Paraclesian transformation potions to top level alchemy perk"

#### Purpose of change

We needed some level 5 alchemy items for Xedra X Bombastic Perks. Since we have the cool new paraclesian line, why not some potions to become one.

#### Describe the solution

Adds a set of Imbuement potions you can get from the Alchemy level 5 perk, which add the starting trait and recipes as if you had started as a paraclesian.

For balance sake, only works if you're not post threshold, so I've added a new condition that checks if you're post threshold.

### CODE REVIEW

File: data/mods/Xedra_Evolved/items/alchemy.json [hunk 1]
```diff
@@ -299,5 +299,227 @@
...
+    "symbol": "!",
+    "color": "light_cyan",
+    "container": "flask_glass",
+    "quench": 2,
+    "healthy": 7,
+    "fun": -30,
+    "flags": [ "NO_INGEST", "WATER_DISSOLVE", "EDIBLE_FROZEN" ],
+    "use_action": {
+      "type": "effect_on_conditions",
+      "description": "Drink the potion. Will not work on those too distant from humanity already.",
```
Comment (Maleclypse) [inline data/mods/Xedra_Evolved/items/alchemy.json:323]:
```suggestion
      "description": "Drink the potion.  Will not work on those too distant from humanity already.",
```

File: data/mods/Xedra_Evolved/items/alchemy.json [hunk 2]
```diff
@@ -299,5 +299,227 @@
...
```
Comment (Maleclypse) [inline data/mods/Xedra_Evolved/items/alchemy.json:360]:
```suggestion
      "description": "Drink the potion.  Will not work on those too distant from humanity already.",
```

File: data/mods/Xedra_Evolved/items/alchemy.json [hunk 3]
```diff
@@ -299,5 +299,227 @@
...
```
Comment (Maleclypse) [inline data/mods/Xedra_Evolved/items/alchemy.json:397]:
```suggestion
      "description": "Drink the potion.  Will not work on those too distant from humanity already.",
```

File: data/mods/Xedra_Evolved/items/alchemy.json [hunk 4]
```diff
@@ -299,5 +299,227 @@
...
```
Comment (Maleclypse) [inline data/mods/Xedra_Evolved/items/alchemy.json:434]:
```suggestion
      "description": "Drink the potion.  Will not work on those too distant from humanity already.",
```

File: data/mods/Xedra_Evolved/items/alchemy.json [hunk 5]
```diff
@@ -299,5 +299,227 @@
...
```
Comment (Maleclypse) [inline data/mods/Xedra_Evolved/items/alchemy.json:471]:
```suggestion
      "description": "Drink the potion.  Will not work on those too distant from humanity already.",
```

File: data/mods/Xedra_Evolved/items/alchemy.json [hunk 6]
```diff
@@ -299,5 +299,227 @@
...
```
Comment (Maleclypse) [inline data/mods/Xedra_Evolved/items/alchemy.json:508]:
```suggestion
      "description": "Drink the potion.  Will not work on those too distant from humanity already.",
```

**Comment by Maleclypse** [comment]
I pushed all the linting fixes before I pulled this over to my local to compile and test.

**Comment by Maleclypse** [comment]
imbuement works

Not sure why it's not working for you, probably needed a fresh world.  But the Salamander imbuement worked fine for me.

**Comment by Maleclypse** [comment]
OK so I've been playing with this and I want the threshold check to work for future things but I tested some things and I think that if someone has crossed a threshold they will be locked out of going post thresh in the elemental tree.  So the biggest issue would be someone taking multiple imbuements.   I think the solution for that is to create a hidden trait `"IMBUED"` that only exists to be checked in the condition and if the PC has that trait from drinking an imbuement potion before then that would prevent the PC from gaining a second.   I guess I don't mind a post thresh ursine or cephalopod getting some low level fire or plant mutations.  It looks like as long as you have a threshhold already it should be impossible to get a second.

**Comment by Maleclypse** [comment]
@RenechCDDA if you have a moment can you look at the has_threshold condition in this PR? It’s not failing tests and builds but it doesn’t seem to be firing in game.

File: src/character.h [hunk 7]
```diff
@@ -1284,6 +1284,10 @@ class Character : public Creature, public visitable
          *  Uses has_opposite_trait(), has_lower_trait(), and has_higher_trait() to determine conflicts.
          */
         bool has_conflicting_trait( const trait_id &flag ) const;
+        /**
+         * Returns true if the player has a threshold mutation
+         */
+        bool has_threshold() const;
```
Comment (Maleclypse) [inline src/character.h:1290]: Is it a problem that has_threshold already exists in character.h and is referring to the aiming code?  

```
 std::string action;
  366      std::string help;
  367:     bool has_threshold;
  368      int threshold;
  369      bool operator==( const aim_type &other ) const {
  370:         return has_threshold == other.has_threshold &&
  371                 threshold == other.threshold &&
  372                 name == other.name &&
```

Comment (MorvarchPrincess) [reply src/character.h:1290]: Shouldn't be, that's just a variable on a different object type

File: src/character.cpp [hunk 8]
```diff
@@ -8363,14 +8363,14 @@ void Character::update_vitamins( const vitamin_id &vit )
     }
     if( lvl > 0 ) {
         if( has_effect( def ) ) {
-            get_effect( def ).set_intensity( lvl, true );
+            get_effect( def ).set_intensity( lvl, is_avatar() );
```
Comment (Maleclypse) [inline src/character.cpp:8366]:
```suggestion
            get_effect( def ).set_intensity( lvl, true );
```

File: src/character.cpp [hunk 9]
```diff
@@ -8363,14 +8363,14 @@ void Character::update_vitamins( const vitamin_id &vit )
...
         } else {
             add_effect( def, 1_turns, true, lvl );
         }
     }
     if( lvl < 0 ) {
         if( has_effect( exc ) ) {
-            get_effect( exc ).set_intensity( -lvl, true );
+            get_effect( exc ).set_intensity( -lvl, is_avatar() );
```
Comment (Maleclypse) [inline src/character.cpp:8373]:
```suggestion
            get_effect( exc ).set_intensity( -lvl, true );
```

File: src/character.cpp [hunk 10]
```diff
@@ -11568,7 +11568,7 @@ const Character *Character::get_book_reader( const item &book,
         return nullptr;
     }
 
-    // Check for conditions that disqualify us only if no NPCs can read to us
+    // Check for conditions that disqualify us only if no other Characters can read to us
```
Comment (Maleclypse) [inline src/character.cpp:11571]:
```suggestion
    // Check for conditions that disqualify us only if no NPCs can read to us
```

File: src/character.cpp [hunk 11]
```diff
@@ -11588,17 +11588,17 @@ const Character *Character::get_book_reader( const item &book,
         return nullptr;
     }
 
-    //Check for NPCs to read for you, negates Illiterate and Far Sighted
-    //The fastest-reading NPC is chosen
+    //Check for other Characters to read for you, negates Illiterate and Far Sighted
```
Comment (Maleclypse) [inline src/character.cpp:11591]:
```suggestion
    //Check for NPCs to read for you, negates Illiterate and Far Sighted
```

File: src/character.cpp [hunk 12]
```diff
@@ -11588,17 +11588,17 @@ const Character *Character::get_book_reader( const item &book,
...
+    //The fastest-reading Character is chosen
```
Comment (Maleclypse) [inline src/character.cpp:11592]:
```suggestion
    //The fastest-reading NPC is chosen
```

File: src/character.cpp [hunk 13]
```diff
@@ -11588,17 +11588,17 @@ const Character *Character::get_book_reader( const item &book,
...
     if( is_deaf() ) {
         reasons.emplace_back( _( "Maybe someone could read that to you, but you're deaf!" ) );
         return nullptr;
     }
 
     time_duration time_taken = time_duration::from_turns( INT_MAX );
-    auto candidates = get_crafting_helpers();
+    std::vector<Character *> candidates = get_crafting_helpers();
```
Comment (Maleclypse) [inline src/character.cpp:11599]:
```suggestion
    auto candidates = get_crafting_helpers();
```

File: src/character.cpp [hunk 14]
```diff
@@ -11588,17 +11588,17 @@ const Character *Character::get_book_reader( const item &book,
...
 
-    for( const npc *elem : candidates ) {
+    for( const Character *elem : candidates ) {
```
Comment (Maleclypse) [inline src/character.cpp:11601]:
```suggestion
    for( const npc *elem : candidates ) {
```

File: src/character.cpp [hunk 15]
```diff
@@ -11847,9 +11847,8 @@ int Character::get_lift_str() const
 int Character::get_lift_assist() const
 {
     int result = 0;
-    const std::vector<npc *> helpers = get_crafting_helpers();
-    for( const npc *np : helpers ) {
-        result += np->get_lift_str();
+    for( const Character *guy : get_crafting_helpers() ) {
+        result += guy->get_lift_str();
```
Comment (Maleclypse) [inline src/character.cpp:11851]:
```suggestion
    const std::vector<npc *> helpers = get_crafting_helpers();
    for( const npc *np : helpers ) {
        result += np->get_lift_str();
```

**Review by Maleclypse — COMMENTED** [review]
to reverse the c++ changes

File: data/mods/Xedra_Evolved/items/alchemy.json [hunk 16]
```diff
@@ -299,5 +299,149 @@
...
```
Comment (Maleclypse) [inline data/mods/Xedra_Evolved/items/alchemy.json:323]:
```suggestion
      "description": "Drink the potion.  Will not work on those too distant from humanity already.",
```

File: data/mods/Xedra_Evolved/items/alchemy.json [hunk 17]
```diff
@@ -299,5 +299,149 @@
...
```
Comment (Maleclypse) [inline data/mods/Xedra_Evolved/items/alchemy.json:347]:
```suggestion
      "description": "Drink the potion.  Will not work on those too distant from humanity already.",
```

File: data/mods/Xedra_Evolved/items/alchemy.json [hunk 18]
```diff
@@ -299,5 +299,149 @@
...
```
Comment (Maleclypse) [inline data/mods/Xedra_Evolved/items/alchemy.json:371]:
```suggestion
      "description": "Drink the potion.  Will not work on those too distant from humanity already.",
```

File: data/mods/Xedra_Evolved/items/alchemy.json [hunk 19]
```diff
@@ -299,5 +299,149 @@
...
```
Comment (Maleclypse) [inline data/mods/Xedra_Evolved/items/alchemy.json:395]:
```suggestion
      "description": "Drink the potion.  Will not work on those too distant from humanity already.",
```

File: data/mods/Xedra_Evolved/items/alchemy.json [hunk 20]
```diff
@@ -299,5 +299,149 @@
...
```
Comment (Maleclypse) [inline data/mods/Xedra_Evolved/items/alchemy.json:419]:
```suggestion
      "description": "Drink the potion.  Will not work on those too distant from humanity already.",
```

File: data/mods/Xedra_Evolved/items/alchemy.json [hunk 21]
```diff
@@ -299,5 +299,149 @@
...
```
Comment (Maleclypse) [inline data/mods/Xedra_Evolved/items/alchemy.json:443]:
```suggestion
      "description": "Drink the potion.  Will not work on those too distant from humanity already.",
```

**Review by Maleclypse — COMMENTED** [review]
Linting

**Comment by RenechCDDA** [comment]
> @RenechCDDA if you have a moment can you look at the has_threshold condition in this PR? It’s not failing tests and builds but it doesn’t seem to be firing in game.

Looks like this was resolved with Andrei8l yesterday, if I read the development discord correctly?

**Comment by MorvarchPrincess** [comment]
@RenechCDDA The initial problem was never solved, but I switched to a different solution.

**Pull request merged**